### PR TITLE
Add support for remote config in voice search

### PR DESCRIPTION
**Pull request by CrisBarreiro**
Task/Issue URL: https://app.asana.com/0/0/1205292240653427/f

Project: https://app.asana.com/0/72649045549333/1205122648906231/f

Depends on https://github.com/duckduckgo/privacy-configuration/pull/1229

### Description
Use remote config to determine whether or not to show voice search

### Steps to test this PR

Sample 
```
    "voiceSearch": {
      "exceptions": [],
      "state": "enabled",
      "settings": {
        "minVersion": 33,
        "excludedManufacturers": [
          {
            "name": "{manufacturerName}"
          }
        ]
      }
    },
```

_Feature enabled, no excluded manufacturers, device running Android 13 -> Voice search icon_
- [x] Run the app on a device with Android 13
- [x] Create a copy of remote config using the sample above
- [x] Remove all values from `excludedManufacturers`
- [x] Open the app 
- [x] Look for a microphone icon on the search bar

_Feature enabled, excluded manufacturers with other values, device running Android 13 -> Voice search icon_
- [x] Run the app on a device with Android 13
- [x] Create a copy of remote config using the sample above
- [x] Add a manufacturer different than the current one to `excludedManufacturers`
- [x] Open the app 
- [x] Look for a microphone icon on the search bar

_Feature enabled, excluded manufacturers with other values, device running Android 12 -> No voice search icon_
- [x] Run the app on a device with Android 12
- [x] Create a copy of remote config using the sample above
- [x] Add a manufacturer different than the current one to `excludedManufacturers`
- [x] Open the app 
- [x] There's no microphone icon on the search bar

_Feature enabled, excluded manufacturers with other values, device running Android 13 -> No voice search icon_
- [x] Run the app on a device with Android 13
- [x] Create a copy of remote config using the sample above
- [x] Add current manufacturer to `excludedManufacturers`
- [x] Open the app 
- [x] There's no microphone icon on the search bar

_Feature disabled, no excluded manufacturers, device running Android 13 -> No voice search icon_
- [x] Run the app on a device with Android 13
- [x] Create a copy of remote config using the sample above
- [x] Remove all values from `excludedManufacturers`
- [x] Open the app 
- [x] There's no microphone icon on the search bar

### CODE REVIEW

**Comment by CrisBarreiro** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3477** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3477" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3477?utm_source=stack-comment).

File: voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/RealVoiceSearchAvailability.kt [hunk 1]
```diff
@@ -16,31 +16,37 @@
...
         private const val LANGUAGE_TAG_ENG_US = "en-US"
         private const val URL_DDG_SERP = "https://duckduckgo.com/?"
     }
 
-    private val allowList = listOf("pixel 6", "pixel 6 pro")
-
     override val isVoiceSearchSupported: Boolean
         get() = configProvider.get().run {
-            hasValidDevice(deviceModel.lowercase()) && hasValidVersion(sdkInt) && isOnDeviceSpeechRecognitionSupported && hasValidLocale(languageTag)
+            voiceSearchFeature.self().isEnabled() &&
```
Comment (anikiki) [inline voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/RealVoiceSearchAvailability.kt:39]: This condition is duplicated.

File: voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/remoteconfig/VoiceSearchFeatureRepository.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    val minVersion: Int?
+}
+
+class RealVoiceSearchFeatureRepository constructor(
+    val database: VoiceSearchDatabase,
+    coroutineScope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider,
+) : VoiceSearchFeatureRepository {
+
+    private val autofillDao: VoiceSearchDao = database.voiceSearchDao()
```
Comment (anikiki) [inline voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/remoteconfig/VoiceSearchFeatureRepository.kt:40]: Rename to `voiceSearchDao`.

File: voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/remoteconfig/VoiceSearchFeatureRepository.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    override fun updateAllExceptions(exceptions: List<Manufacturer>, minVersion: Int) {
+        autofillDao.updateAll(exceptions.map { ManufacturerEntity(it.name) }, MinVersionEntity(minVersion))
+        loadToMemory()
+    }
+
+    private fun loadToMemory() {
+        manufacturers.clear()
+        autofillDao.getAllManufacturers().map { manufacturers.add(Manufacturer(it.name)) }
```
Comment (anikiki) [inline voice-search/voice-search-impl/src/main/java/com/duckduckgo/voice/impl/remoteconfig/VoiceSearchFeatureRepository.kt:57]: Adding each manufacturer is more expensive than adding all of them when working with  `CopyOnWriteArrayList` as a new collection is created for every element added. I don't know how many manufacturers we'll have.

Can do:
```
val manufacturerEntities = autofillDao.getAllManufacturers()
manufacturers.addAll(manufacturerEntities.map { Manufacturer(it.name) })
```

**Review by anikiki — APPROVED** [review]
Left a few comments.
Looks good and works as expected! :tada:

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 4]
```diff
@@ -2875,6 +2875,12 @@ class BrowserTabViewModel @Inject constructor(
         }
     }
 
+    fun voiceSearchDisabled() {
+        omnibarViewState.value = currentOmnibarViewState().copy(
```
Comment (karlenDimla) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:2879]: I think this is not evaluated right away. Consider this steps:
1.) Install Build
2.) Search for something and verify you are in DDG search results page.
3.) Verify that microphone is visible
4.) Hide icon via settings
5.) Return to browser

After this the icon is still visible and only disappears after clicking on a link, opening a new tab etc.

File: app/src/main/java/com/duckduckgo/widget/WidgetVoiceSearchStatusListener.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+package com.duckduckgo.widget
+
+import android.content.Context
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.voice.api.VoiceSearchStatusListener
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+@ContributesBinding(AppScope::class)
+class WidgetVoiceSearchStatusListener @Inject constructor(
```
Comment (karlenDimla) [inline app/src/main/java/com/duckduckgo/widget/WidgetVoiceSearchStatusListener.kt:26]: Where is this used?

File: app/src/main/java/com/duckduckgo/widget/VoiceSearchWidgetConfigurator.kt [hunk 6]
```diff
@@ -33,7 +33,7 @@ class VoiceSearchWidgetConfigurator @Inject constructor(
         remoteViews: RemoteViews,
         fromFavWidget: Boolean,
     ) {
-        if (voiceSearchAvailability.isVoiceSearchSupported) {
+        if (voiceSearchAvailability.isVoiceSearchAvailable) {
```
Comment (karlenDimla) [inline app/src/main/java/com/duckduckgo/widget/VoiceSearchWidgetConfigurator.kt:36]: Not in this specific part of code but seems like the widgets are not being updated  when remote config is updated.

Consider these steps:
1.) Install build
2.) Verify that icon is visible
3.) Add widget to home and verify icon is visible there.
4.) Update remote config to exclude device’s manufacturer
5.) Restart app
6- Verify that icon disappears from browser

Howere when I check the icon, it is not updated.

File: app/src/main/java/com/duckduckgo/widget/WidgetVoiceSearchStatusListener.kt (hunk 5 above)
Comment (karlenDimla) [reply app/src/main/java/com/duckduckgo/widget/WidgetVoiceSearchStatusListener.kt:26]: Ah nvm, found it.

**Review by anikiki — APPROVED** [review]
LGTM :shipit:

**Pull request merged**